Adding new version of Minuit2 standalone with CMake build

henryiii opened the pull request:
This adds a new standalone Minuit2 build system built in CMake to replace the old one that no longer works. The following features are part of the design:

* Full featured CMake build, based on https://github.com/GooFit/Minuit2
    - Only requires CMake 3.1+
    - Support for adding as a subdirectory, using from build directory, or installing
    - Works inside of ROOT or standalone
    - Support for CPack binary `make package`
    - Options for MPI and OpenMP integrated with main ROOT build
    - Support for macOS, Linux, and Windows
    - Version captured from ROOT
    - CTest support
    - `make docs` support (CMake 3.8+ only)
* No need for extra "prepare" shell scripts, everything is done cross-platform through CMake and CPack
    - Files can be copied in to make redistributable source with `-Dminuit2-standalone=ON`
    - `make package_source` on Unix creates `.tar.gz` and `.zip` files (non-binary CPack)
    - `make purge` can remove copied files
* Helpful `README.md` and `DEVELOP.md` files

### To produce a source distribution:
Make sure you have CMake 3.1+ (ROOT requires 3.4.3+, so that's likely). Then, from `/math/minuit2/build` run:

```bash
cmake .. -Dminuit2-standalone=ON
make package_source
make purge
```

### You can also build Minuit2 for testing (standalone option on or off) with:

```
make
```

### And you can make a prebuilt binary package with:

```
make package
```

(Standard CPack option for different generators apply.)

### Maintenence

If new files are needed by Minuit2 due to additions to ROOT, they should be added to the source files lists in `/math/minuit2/src/Math/CMakeLists.txt` and `/math/minuit2/src/CMakeLists.txt` (depending on if it's a new Math or Minuit2 requirement).


### Windows
This works on Windows; and you can even then create binary installers for Windows trivially. (Though "package_source" does not seem to be supported on Windows). This is the procedure on Windows:

First, install Visual Studio 17, Git, CMake, and NSIS. Then from the `math/minuit2/build directory` of your ROOT checkout, run:

```
"C:\Program Files\CMake\bin\cmake.exe" ..
"C:\Program Files\CMake\bin\cmake.exe" --build .
"C:\Program Files\CMake\bin\cmake.exe" --build . --target package
```

This creates a `.exe` installer file.

amadio commented:
@phsft-bot build

henryiii commented:
Looks like the Windows build failed due to a git clone issue.

bellenot commented:
This was the issue:
 `stderr: fatal: Unable to create 'C:/build/workspace/root-pullrequests-build/root/.git/index.lock': File exists.`
This should be fixed now.

henryiii commented:
@phsft-bot build just on windows10/vc15

amadio commented:
@phsft-bot build just on windows10/vc15

amadio commented:
@henryiii I think only admins can trigger builds...

henryiii commented:
Thought I'd try, and also made it easier for an admin to copy and paste. :)

amadio commented:
BTW, if this does not include the commit I merged from the other PR, then it may fail. In that case, please rebase on the current master branch and push again, please.

henryiii commented:
I believe it builds the merged result; it passed.

Is it possible to access build artifacts from Jenkins? We could add the standalone generation to Jenkins and make source and binary dists for several systems. (In a future PR).

henryiii commented:
I rebased to get the Windows fix locally. Here are some files for testing:
*  [Minuit2-6.13.1-Source.tar.gz](https://github.com/root-project/root/files/1770290/Minuit2-6.13.1-Source.tar.gz)
* [Minuit2-6.13.1-Source.zip](https://github.com/root-project/root/files/1770292/Minuit2-6.13.1-Source.zip)
* [Minuit2-6.13.1-Darwin.tar.gz](https://github.com/root-project/root/files/1770293/Minuit2-6.13.1-Darwin.tar.gz)
* [Minuit2-6.13.1-Linux.tar.gz](https://github.com/root-project/root/files/1770294/Minuit2-6.13.1-Linux.tar.gz)

(GitHub doesn't support `.sh` files, so not uploading the self-unpacking installers)


(Linux packages created with: `docker run -v (pwd):/root/root -it alpine`, making and changing to a `build-linux` directory, running `apk --no-cache add g++ make cmake` then `cmake ..` followed by `make package`) I can't upload a Windows package right now because I have to be at CERN to access my OpenStack instance.

amadio commented:
> I can't upload a Windows package right now because I have to be at CERN to access my OpenStack instance.

You can always ssh into lxplus and log into your CERN machines from there.

amadio commented:
@phsft-bot build

henryiii commented on math/minuit2/standalone/CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+
+    # For CMake < 3.9, we need to make the target ourselves
+    if(NOT TARGET OpenMP::OpenMP_CXX)
+        add_library(OpenMP_TARGET INTERFACE)
```
It's much uglier and harder to build an IMPORTED INTERFACE library in CMake until the upcoming CMake 3.11, and this generally works well. I can change it, though.

amadio replied:
If it's too hard, don't bother. We should only make sure that it's not exported if it's not needed.

henryiii commented on math/minuit2/standalone/.gitignore:
```diff
@@ -0,0 +1,5 @@
+*.h
+version_number
+LICENSE
+LGPL2_1.txt
```
Where should I put the ignore lines, in the master .gitignore?

amadio replied:
Well, I think that if we don't copy things, but instead share the same CMakeLists.txt with an `if` condition for the standalone case, then the files can go away. The reason to ignore them is that they are copied, right?

henryiii commented:
Some headers and source files are needed from Math, and a header from Fit. I could copy them in as part of the build like I do now with the standalone dir? They have to be in the minuit directory, since they need to be available if ROOT is not there.

henryiii commented on math/minuit2/standalone/CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+
+
+# Inherit default from parent project if not main project
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```
No, building as part of a project that is not ROOT is supported, using `add_subdirectory(Minuit2)`.

amadio commented:
Cool, I think that we can move the stuff that is needed from TMath into Minuit2, and just add a compatibility header in TMath that includes the one in Minuit2 so that nothing breaks.

henryiii commented:
But what about if Math becomes available standalone someday? Or another part of ROOT becomes standalone that needs Math? Since there's work being done on making ROOT more modular, I don't think it's a good idea make Math something built into Minuit. I think making `inc/Math`, `src/Math`, and `inc/Fit` and just picking out the used parts (that about 20 files, I believe) is a reasonable in-between solution.

amadio commented:
It has to be looked case by case. Another option is to make Minuit2 not depend on the headers from Math at all. I will take a look just in case.

amadio commented:
Yeah, it looks like Minuit2 and Math are quite entangled at the moment, unfortunately.

henryiii commented:
I think the copy on build idea will work for now, and this could be improved in the future.

amadio commented:
Ok, but please copy headers to `${CMAKE_BUILD_DIR}/include`, like the rest of ROOT headers, since that eliminates the need for gitignore files. How do you usually build it standalone? Do you clone ROOT or only Minuit2 somehow?

henryiii commented:
Originally, I used https://github.com/GooFit/Minuit2 - that uses a python script to re-copy the changed files from the ROOT sources. Being a git repository, it also easily is included as a submodule in the `/extern` folder for GooFit, for example.

After meeting and talking with @lmoneta a couple of days ago over a coffee, he outlined the procedure I initially implemented. He showed me the old "build" directory, with an old bash script that copied the necessary files out. He wanted me to make a new folder to eventually replace his build folder. The idea is that you could create source distributions that would be packaged and made available as downloads to continue the outdated ones available here: http://project-mathlibs.web.cern.ch/project-mathlibs/minuit/release/download.html. I added this though `make package_source`, and included `make package` for binary distributions as well.

The integrated version does look a bit better, though it is a bit more invasive and not nearly as well separated. I do need to copy to `/inc/Math`, etc, since a source distribution needs to be in the source directory. Everything is copied to `INSTALL_PREFIX/include` when it's installed or packaged for a binary distribution.

As for use, a git only solution really is not feasible, unfortunately, with any method that does not include git submodules. The current methods for a user are:

* Install beforehand from one of the downloadable packages, either building a source package or installing a binary package. This is the way some people will get it, but not ideal for a modern package that wants to have minimal dependencies.
* Have CMake download and build the package with `ExternalProject` - probably the best solution for now, still not as nice for something like GooFit that already has all dependencies as git submodules.
* Check out all of ROOT as a submodule, then `add_subdirectory(extern/root/math/minuit2)`. Much more disk space usage that needed, but otherwise also a decent solution. Might even be best.

All three methods should be supported by this standalone build system.

henryiii commented:
The CMake build is now much more heavily integrated with the folder structure. This has the positive side that less copying is going on, and the downside that it has to fill in the source directory with the items it needs from ROOT (they are ignored and easy to clean with `make purge`, though). Like before, you can build in place or make a redistributable source package.

Only added one `.gitignore`. Haven't really touched the ROOT Minuit2 build system, though some fixes would be good there too, but that's a later PR.

(Instructions have not been updated to new version yet, and not as well tested yet)

I can revert to the old system if that's better. The original reason for using CMake to do the copy also was because that avoided having to write a `.sh` file, and was more portable.

amadio commented on math/minuit2/StandAlone.cmake:
```diff
@@ -12,8 +12,8 @@ endif()
-copy_standalone(../../../build . version_number)
-copy_standalone(../../.. . LICENSE LGPL2_1.txt)
+copy_standalone(../../build . version_number)
+copy_standalone(../.. . LICENSE LGPL2_1.txt)
```
What's wrong with `${CMAKE_COMMAND} -E copy [src] [dst]` or even `file(COPY ...)`?

henryiii replied:
This one makes sure the target directory exists, adds it to the purge target, and most importantly does not make the copy unless this is inside ROOT. This is what allows this to work when it is actually "standalone".

henryiii replied:
Internally it uses `configure_file(... COPYONLY)`, but it could use either of those two as well.

henryiii commented:
I can't test MPI locally, since I've got something based on MPI3, which has removed the C++ bindings. I'd need #1689 to test MPI.

amadio commented:
I will test here.

henryiii commented:
Should I have it build on either Travis or Jenkins? (I don't know where to add it to Jenkins). Even if the artifacts are not available, it would be nice to make sure it still builds if there are changes to Minuit2.

amadio commented:
ROOT is a bit too heavy for Travis, and Jenkins is managed outside the repository. Just let me know which options the build should enable, and I will trigger a build for you.

henryiii commented:
It would only build Minuit2, not all of ROOT, so that seems possible for Travis.

amadio commented:
@henryiii I have made changes to the `CMakeLists.txt` of Minuit2 to add the options to build with OpenMP and MPI support. I am not sure that all the stuff to manage Minuit2 outside of ROOT belongs in the ROOT repo, however. This is especially true for all the copying of files in the source directory that needs to happen. It may be better to maintain that part in the fork.

pcanal commented:
In general, I am uneasy with solution that modify the source tree, they tend to be hard to debug/track when things go wrong (in particular when switching the source tree from one git branch to the other, etc..) 

Wouldn't be it be possible to copy the necessary files into the build directory instead?

henryiii commented:
I really prefer a separate repo, exactly the design of `GooFit/Minut2` (though moved to `root-project/minuit2`). It has the nice feature that you can use it on multiple ROOT versions (producing a 6.10 version, for example), it keeps ROOT clean, and is clearly seperate. (I'm calling that Version 1)

The first separate folder (Version 2) method was from a discussion with @lmoneta, and the merged folder method (Version 3) was from a request by @amadio. When I get back to CERN (Wednesday), can we schedule another meeting with interested parties to set up a plan? We need to pick one of the three options and focus (my) effort on that.

And, no, since you need to make a distributable *source* package, you can't copy all files to the build folder. You can copy all files to a `/standalone` folder, which is what Version 2 did.

pcanal commented:
> And, no, since you need to make a distributable source package, you can't copy all files to the build > folder. You can copy all files to a /standalone folder

I would think that preparing the modified version of the source into a subdirectory of build would work as well .... [As an aside, I am not sure "distributable modified source package" is the best solution, given the current bandwidth and disk space, just grabbing 'root.git' and building from there a standalone minuit2 library might work as well ...]

Axel-Naumann commented:
In principle I agree with Philippe: the cost of checking out root.git and building only the minuit2 subset is minimal. But we're not going to make a deal breaker out of this: as long as our (esp maintenance) cost of integrating your approach is minimal we should just merge it. Guilherme will guide you through the implementation issues should there be any left.

henryiii commented:
I've modified the system to require an explicit request to modify the source tree to build the standalone redistributable; it now builds in place with no copying if you just build inside the ROOT source tree. I've added a Travis test for the build (both copy and no copy versions) since the build is very fast.

So the current version allows:
* Building directly from the `/math/minuit2` directory while inside ROOT. No copies or modifications to the source directory are made. `add_subdirectory`, `make install`, `make package` (binary), and `find_package`
* Building directly from the `/math/minuit2` directory with `-Dminuit2-standalone=ON` set. This will copy the needed files into the minuit2 source directory so that `make package_source` is supported - allowing stand-alone redistributable packages to be produced. `make purge` cleans the source directory.
* Building from stand-alone redistributable packages produced above. Using `add_subdirectory`, `make install`, `make package` (binary), and `find_package` are all supported.

I think this is flexible enough to cover the possible uses while not adding too much maintenance.  The `README.md` and `DEVELOP.md` files have been updated, and the CMake files have fairly extensive comments.

henryiii commented:
Any updates? Does anything need changing? I think it's best to focus on this one before #1689, since that one will be easier to develop and test once this is in and standalone Minuit2 builds are possible.

henryiii commented:
Fixed the Travis build, moved build script, and added the cmake build-and-test test to the tests (Turns out, CMake's own source has lots of examples of that since that's what a test of CMake itself is generally doing). Fixed a bug in the calculation of the standalone directory, and simplified the CMake code a bit. `make package_source` now produces an error if in used in the wrong configuration.

I've tried to avoid making any changes to the parts of Travis that are not affected, but there's a bug in the yaml: making a node anchor is also a normal definition; so if no packages are added, they get the default (which includes a non-existent timeout package).

I'll see if I can make a separate PR with some improvements for Travis clarity/speed not related to Minuit2.

This PR should be ready for review.

amadio commented:
@henryiii Thank you! I will have another look at the changes and merge when I'm done.

amadio commented:
@phsft-bot build

lmoneta commented:
Thank you very much Harry for this very useful contribution!
It is now merged in the master
